### DISCUSSION TITLE
Creating two tables in a database from two csv files

### DISCUSSION THREAD

**Discussion by ag-din** [sqlalchemy/sqlalchemy#8928] [Usage Questions] [answered]
Hi everyone!

I am a beginner with SQLAlchemy, and I want to create a SQLite database with 2 tables from 2 csv files ([fileA.csv](https://github.com/sqlalchemy/sqlalchemy/files/10149451/fileA.csv) and [fileB.csv](https://github.com/sqlalchemy/sqlalchemy/files/10149452/fileB.csv)). The file fileA.csv consists of 20 columns (labelled A1, A2, ..., A20 respectively) and has 100 rows.  The file fileB.csv consists of 50 columns (labelled B1, B2, ..., B50 respectively) and has 200 rows. The data type of each column is float values. 

The code I apply to create this database with two tables is:

```
# Libraries
import json
import pandas as pd
from sqlalchemy import create_engine, MetaData, Table, insert, select
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Float
from sqlalchemy.orm import sessionmaker


# Create a SQLite database with 2 tables

# Create a base class
Base = declarative_base()

# Generate an Engine instance towards SQLite
engine = create_engine("sqlite:///DB.db", echo=False)

#Create the session
conn = engine.connect()

# Define a mapped class from base class
class Template(Base):
    
    __abstract__ = True
    
    def __init__(self, name: str, num_cols: int):
        self.name = name
        self.num_cols = num_cols
        __table__ = Table(
                          name,
                          Base.metadata,
                          *[Column(name + str(i+1), Float) for i in range(num_cols)]
                        )

# Define table structures (I have in mind create multiple tables, 
# so I thought that define a class Template could be useful)

A_table = Template(name='A', num_cols=20)
B_table = Template(name='B', num_cols=50)

# Create tables
Template.metadata.create_all(engine)

# Add data
A_json = json.loads(pd.read_csv('fileA.csv').to_json(orient='records'))
B_json = json.loads(pd.read_csv('fileB.csv').to_json(orient='records'))

conn.execute(Template.metadata.tables['A'].insert(), A_json)
conn.execute(Template.metadata.tables['B'].insert(), B_json)
```

On the other hand, I want to define a function that accesses one of the tables and do a query (read data from a column of a table). I did this:

```
def get_column_from_table(engine, matadata, table_name, column_id:str):
    f = None
    
    if table_name == 'A':
        table = Template.metadata.tables[table_name]
        query = select([table.c[table_name+column_id]])
        f = engine.execute(query).fetchall()

    elif table_name == 'B':
        table = Template.metadata.tables[table_name]
        query = select([table.c[table_name+column_id]])
        f = engine.execute(query).fetchall()

    return f
```

**Questions:**
1. Is this approach that I have done to load two tables from cvs files into a database correct? On the other hand, I am strongly interested in knowing if it is possible to make a more optimal code to do what I have mentioned.
2. Regarding the function that reads the values of a column of one of the tables, is it correct? It is not clear to me through which objects and how I can access a table and do a query.

Thank you!

**Comment by CaselIT** [comment, marked as answer]
> 
> 1. Is this approach that I have done to load two tables from cvs files into a database correct? On the other hand, I am strongly interested in knowing if it is possible to make a more optimal code to do what I have mentioned.

You can do without the base, that's used only when using the ORM
```py
# Create a base class
Base = declarative_base()

# Define a mapped class from base class
class Template(Base):
    
    __abstract__ = True
    
    def __init__(self, name: str, num_cols: int):
        self.name = name
        self.num_cols = num_cols
        __table__ = Table(
                          name,
                          Base.metadata,
                          *[Column(name + str(i+1), Float) for i in range(num_cols)]
                        )
```
you can do
```py
metadata = MetaData()
class Template:
    def __init__(self, name: str, num_cols: int):
        self.name = name
        self.num_cols = num_cols
        self.table = Table(
                          name,
                          metadata,
                          *[Column(name + str(i+1), Float) for i in range(num_cols)]
                        )

A_table = Template(name='A', num_cols=20)
B_table = Template(name='B', num_cols=50)

# or just use a function
def make_table(name: str, num_cols: int):
        return Table(
                          name,
                          metadata,
                          *[Column(name + str(i+1), Float) for i in range(num_cols)]
                        )

A_table = make_table(name='A', num_cols=20)
B_table = make_table(name='B', num_cols=50)
```
> 
> 2. Regarding the function that reads the values of a column of one of the tables, is it correct? It is not clear to me through which objects and how I can access a table and do a query.
> 

The dict `metadata.tables[table_name]` contains the tables you defined with the metadata, so you can access them using it.
The function seems to be mostly ok

**Reply by ag-din** [reply]
Thank you for your answer and time.
